projects/common/src/telemetry/telemetry.ts

```diff
@@ -0,0 +1,43 @@
+import { InjectionToken } from '@angular/core';
+import { Dictionary } from './../utilities/types/types';
+
+export interface UserTelemetry {
+  readonly config: UserTelemetryConfig;
+  readonly initialized: boolean;
+  initialize(): void;
```

Comment:
we can. I was thinking they can do a no-op. But optional makes it cleaner

Reply:
one side effect of interface is that required properties like config and initialized have to be public. This increases the coupling.

Reply:
interfaces always reduce coupling. Take initialized as an example - who's the consumer of that? If it's the telemetry orchestrator (what you called the internal one), we don't need it on the interface, we can just track it locally. If it's the specific telemetry provider, it doesn't need to expose it via interface. It's up to the impl to decide whether or not it wants to track it which it can do when its initialize call is made. Instead think of this interface as answering two questions that are key to the relationship between one provider and the orchestrator (and which must be public by their nature):
- what actions does the orchestrator need to tell that provider to take?
- what state does the orchestrator need to know about the provider that it couldn't track itself (which I believe those two readonly props fall under)

Reply:
what about `config`?

Reply:
You wouldn't want it to change or meddle with it after the provider has been initialized. As a public property, this becomes difficult to enforce.

Reply:
the provider doesn't need to expose its config, it only needs to read it in. So given the example I had below:
```ts
UserTelemetryModule.withProviders([
  {
    telemetryProvider: FullStoryTelemetryProvider, // Instead of: UserTelemetryType.FullStory,
    telemetryProviderConfig: { orgId: environment.fullStoryOrgId },
    enablePageTracking: true,
    enableEventTracking: true,
    enableErrorTracking: true
  }
]);

interface ConfiguredUserTelemetryProvider<TConfig> {
    telemetryProvider: UserTelemetryProvider<TConfig>;
    telemetryProviderConfig: TConfig;
    enablePageTracking: boolean;
    enableEventTracking: boolean;
    enableErrorTracking: boolean;
}
```

`FullStoryTelemetryProvider` might implement `UserTelemetryProvider` which perhaps looks like:
```ts
export interface UserTelemetryProvider<TConfig> {
  initialize(config: TConfig): void;
  identify(userTraits: UserTraits): void;
  trackEvent(name: string, eventData: Dictionary<unknown>): void;
  trackPage(url: string): void;
  shutdown(): void;
}
```